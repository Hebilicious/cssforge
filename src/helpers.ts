/**
 * Validates a name to ensure it doesn't contain invalid characters.
 * Throws an error if the name is invalid.
 * @example
 * ```ts
 * validateName("valid-name"); // returns true
 * validateName("invalid.name"); // throws error
 * ```
 */
export function validateName(name: string): boolean {
  if (name.length === 0 || name.includes(".") || name === "value") {
    throw new Error(
      `Invalid name: ${name}. Names must be non-empty strings that do not contain periods.`,
    );
  }
  return true;
}

/**
 * Converts a pixel value to a rem value.
 * If the value is not in pixels, it returns the original value.
 * @example
 * ```ts
 * pxToRem({ value: "16px" }); // "1rem"
 * pxToRem({ value: "1rem" }); // "1rem"
 * pxToRem({ value: "32px", rem: 16 }); // "2rem"
 * ```
 */
export function pxToRem({ value, rem = 16 }: { value: string; rem?: number }): string {
  const pxMatch = value.endsWith("px");
  if (pxMatch) {
    const pxValue = parseFloat(value.slice(0, -2));
    const remValue = pxValue / rem;
    return `${remValue}rem`;
  }
  return value;
}
