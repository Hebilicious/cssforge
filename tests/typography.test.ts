import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { defineConfig, processTypography } from "../src/mod.ts";
import { getLines } from "./helpers.ts";

Deno.test("processTypography - generates correct CSS variables", async (t) => {
  const config = defineConfig({
    typography: {
      fluid: {
        arial: {
          value: {
            minWidth: 320,
            minFontSize: 14,
            minTypeScale: 1.25,
            maxWidth: 1435,
            maxFontSize: 16,
            maxTypeScale: 1.25,
            positiveSteps: 5,
            negativeSteps: 3,
          },
        },
      },
    },
  });

  const expectedSizes = [
    "4xl",
    "3xl",
    "2xl",
    "xl",
    "l",
    "m",
    "s",
    "xs",
    "2xs",
  ];
  const result = processTypography(config.typography);
  const lines = getLines(result.css.root);
  assertEquals(lines.length, expectedSizes.length);

  // Test a specific value for precision
  const xlLine = lines.find((line) => line.includes("--typography_fluid-arial-xl:"));
  assertEquals(
    xlLine,
    "--typography_fluid-arial-xl: clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem);",
  );

  assertEquals(
    Array.from(result.resolveMap.keys()),
    expectedSizes.map((size) => `typography_fluid.arial@${size}`),
  );

  // Test that we have all the expected size variables
  expectedSizes.forEach((size) => {
    const hasSize = lines.some((line) =>
      line.includes(`--typography_fluid-arial-${size}:`)
    );
    assertEquals(hasSize, true, `Missing size ${size}`);
  });
  await assertSnapshot(t, result.css);
  await assertSnapshot(t, Array.from(result.resolveMap.entries()));
});

Deno.test("typography - can handle custom labels and prefixes", async (t) => {
  const config = defineConfig({
    typography: {
      fluid: {
        arial: {
          value: {
            minWidth: 320,
            minFontSize: 14,
            minTypeScale: 1.25,
            maxWidth: 1435,
            maxFontSize: 16,
            maxTypeScale: 1.25,
            positiveSteps: 7,
            negativeSteps: 2,
            prefix: "text",
          },
          settings: {
            customLabel: {
              "-2": "xs",
              "-1": "s",
              "0": "m",
              "1": "l",
              "2": "xl",
              "3": "xxl",
              "4": "h4",
              "5": "h3",
              "6": "h2",
              "7": "h1",
            },
          },
        },
      },
    },
  });

  const expectedSizes = [
    "xs",
    "s",
    "m",
    "l",
    "xl",
    "xxl",
    "h4",
    "h3",
    "h2",
    "h1",
  ];
  const result = processTypography(config.typography);
  const lines = getLines(result.css.root);
  assertEquals(lines.length, expectedSizes.length);

  // Test a specific value for precision
  const xlLine = lines.find((line) => line.includes("--typography_fluid-arial-text-xl:"));
  assertEquals(
    xlLine,
    "--typography_fluid-arial-text-xl: clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem);",
  );

  assertEquals(
    Array.from(result.resolveMap.keys()),
    expectedSizes.toReversed().map((size) => `typography_fluid.arial@${size}`),
  );

  // Test that we have all the expected size variables
  expectedSizes.forEach((size) => {
    const hasSize = lines.some((line) =>
      line.includes(`--typography_fluid-arial-text-${size}:`)
    );
    assertEquals(hasSize, true, `Missing size ${size}`);
  });
  await assertSnapshot(t, result.css);
  await assertSnapshot(t, Array.from(result.resolveMap.entries()));
});

Deno.test("processTypography - can process weights", async (t) => {
  const positiveSteps = 3;
  const negativeSteps = 3;
  const config = defineConfig({
    typography: {
      fluid: {
        arial: {
          value: {
            minWidth: 320,
            minFontSize: 14,
            minTypeScale: 1.25,
            maxWidth: 1435,
            maxFontSize: 16,
            maxTypeScale: 1.25,
            positiveSteps,
            negativeSteps,
          },
        },
      },
      weight: {
        arial: { value: { regular: "500" } },
      },
    },
  });

  const result = processTypography(config.typography);
  const lines = getLines(result.css.root);
  assertEquals(lines.length, positiveSteps + negativeSteps + 1 + 1); // 1 for base size and 1 for weight

  assertEquals(
    Array.from(result.resolveMap.keys()),
    [
      "typography_fluid.arial@2xl",
      "typography_fluid.arial@xl",
      "typography_fluid.arial@l",
      "typography_fluid.arial@m",
      "typography_fluid.arial@s",
      "typography_fluid.arial@xs",
      "typography_fluid.arial@2xs",
      "typography.weight.arial.regular",
    ],
  );
  // Test that we have the weight variable
  const xlLine = lines.find((line) =>
    line.includes("--typography-weight-arial-regular:")
  );
  assertEquals(xlLine, "--typography-weight-arial-regular: 500;");
  await assertSnapshot(t, result.css);
  await assertSnapshot(t, Array.from(result.resolveMap.entries()));
});
