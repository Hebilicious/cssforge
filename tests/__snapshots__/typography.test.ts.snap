export const snapshot = {};

snapshot[`processTypography - generates correct CSS variables 1`] = `
{
  root: "--typography_fluid-arial-4xl: clamp(2.6703rem, 2.5608rem + 0.5474vw, 3.0518rem);
--typography_fluid-arial-3xl: clamp(2.1362rem, 2.0486rem + 0.4379vw, 2.4414rem);
--typography_fluid-arial-2xl: clamp(1.709rem, 1.6389rem + 0.3503vw, 1.9531rem);
--typography_fluid-arial-xl: clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem);
--typography_fluid-arial-l: clamp(1.0938rem, 1.0489rem + 0.2242vw, 1.25rem);
--typography_fluid-arial-m: clamp(0.875rem, 0.8391rem + 0.1794vw, 1rem);
--typography_fluid-arial-s: clamp(0.7rem, 0.6713rem + 0.1435vw, 0.8rem);
--typography_fluid-arial-xs: clamp(0.56rem, 0.537rem + 0.1148vw, 0.64rem);
--typography_fluid-arial-2xs: clamp(0.448rem, 0.4296rem + 0.0918vw, 0.512rem);",
}
`;

snapshot[`processTypography - generates correct CSS variables 2`] = `
[
  [
    "typography_fluid.arial@4xl",
    {
      key: "--typography_fluid-arial-4xl",
      value: "clamp(2.6703rem, 2.5608rem + 0.5474vw, 3.0518rem)",
      variable: "--typography_fluid-arial-4xl: clamp(2.6703rem, 2.5608rem + 0.5474vw, 3.0518rem);",
    },
  ],
  [
    "typography_fluid.arial@3xl",
    {
      key: "--typography_fluid-arial-3xl",
      value: "clamp(2.1362rem, 2.0486rem + 0.4379vw, 2.4414rem)",
      variable: "--typography_fluid-arial-3xl: clamp(2.1362rem, 2.0486rem + 0.4379vw, 2.4414rem);",
    },
  ],
  [
    "typography_fluid.arial@2xl",
    {
      key: "--typography_fluid-arial-2xl",
      value: "clamp(1.709rem, 1.6389rem + 0.3503vw, 1.9531rem)",
      variable: "--typography_fluid-arial-2xl: clamp(1.709rem, 1.6389rem + 0.3503vw, 1.9531rem);",
    },
  ],
  [
    "typography_fluid.arial@xl",
    {
      key: "--typography_fluid-arial-xl",
      value: "clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem)",
      variable: "--typography_fluid-arial-xl: clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem);",
    },
  ],
  [
    "typography_fluid.arial@l",
    {
      key: "--typography_fluid-arial-l",
      value: "clamp(1.0938rem, 1.0489rem + 0.2242vw, 1.25rem)",
      variable: "--typography_fluid-arial-l: clamp(1.0938rem, 1.0489rem + 0.2242vw, 1.25rem);",
    },
  ],
  [
    "typography_fluid.arial@m",
    {
      key: "--typography_fluid-arial-m",
      value: "clamp(0.875rem, 0.8391rem + 0.1794vw, 1rem)",
      variable: "--typography_fluid-arial-m: clamp(0.875rem, 0.8391rem + 0.1794vw, 1rem);",
    },
  ],
  [
    "typography_fluid.arial@s",
    {
      key: "--typography_fluid-arial-s",
      value: "clamp(0.7rem, 0.6713rem + 0.1435vw, 0.8rem)",
      variable: "--typography_fluid-arial-s: clamp(0.7rem, 0.6713rem + 0.1435vw, 0.8rem);",
    },
  ],
  [
    "typography_fluid.arial@xs",
    {
      key: "--typography_fluid-arial-xs",
      value: "clamp(0.56rem, 0.537rem + 0.1148vw, 0.64rem)",
      variable: "--typography_fluid-arial-xs: clamp(0.56rem, 0.537rem + 0.1148vw, 0.64rem);",
    },
  ],
  [
    "typography_fluid.arial@2xs",
    {
      key: "--typography_fluid-arial-2xs",
      value: "clamp(0.448rem, 0.4296rem + 0.0918vw, 0.512rem)",
      variable: "--typography_fluid-arial-2xs: clamp(0.448rem, 0.4296rem + 0.0918vw, 0.512rem);",
    },
  ],
]
`;

snapshot[`typography - can handle custom labels and prefixes 1`] = `
{
  root: "--typography_fluid-arial-text-h1: clamp(4.1723rem, 4.0013rem + 0.8553vw, 4.7684rem);
--typography_fluid-arial-text-h2: clamp(3.3379rem, 3.201rem + 0.6843vw, 3.8147rem);
--typography_fluid-arial-text-h3: clamp(2.6703rem, 2.5608rem + 0.5474vw, 3.0518rem);
--typography_fluid-arial-text-h4: clamp(2.1362rem, 2.0486rem + 0.4379vw, 2.4414rem);
--typography_fluid-arial-text-xxl: clamp(1.709rem, 1.6389rem + 0.3503vw, 1.9531rem);
--typography_fluid-arial-text-xl: clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem);
--typography_fluid-arial-text-l: clamp(1.0938rem, 1.0489rem + 0.2242vw, 1.25rem);
--typography_fluid-arial-text-m: clamp(0.875rem, 0.8391rem + 0.1794vw, 1rem);
--typography_fluid-arial-text-s: clamp(0.7rem, 0.6713rem + 0.1435vw, 0.8rem);
--typography_fluid-arial-text-xs: clamp(0.56rem, 0.537rem + 0.1148vw, 0.64rem);",
}
`;

snapshot[`typography - can handle custom labels and prefixes 2`] = `
[
  [
    "typography_fluid.arial@h1",
    {
      key: "--typography_fluid-arial-text-h1",
      value: "clamp(4.1723rem, 4.0013rem + 0.8553vw, 4.7684rem)",
      variable: "--typography_fluid-arial-text-h1: clamp(4.1723rem, 4.0013rem + 0.8553vw, 4.7684rem);",
    },
  ],
  [
    "typography_fluid.arial@h2",
    {
      key: "--typography_fluid-arial-text-h2",
      value: "clamp(3.3379rem, 3.201rem + 0.6843vw, 3.8147rem)",
      variable: "--typography_fluid-arial-text-h2: clamp(3.3379rem, 3.201rem + 0.6843vw, 3.8147rem);",
    },
  ],
  [
    "typography_fluid.arial@h3",
    {
      key: "--typography_fluid-arial-text-h3",
      value: "clamp(2.6703rem, 2.5608rem + 0.5474vw, 3.0518rem)",
      variable: "--typography_fluid-arial-text-h3: clamp(2.6703rem, 2.5608rem + 0.5474vw, 3.0518rem);",
    },
  ],
  [
    "typography_fluid.arial@h4",
    {
      key: "--typography_fluid-arial-text-h4",
      value: "clamp(2.1362rem, 2.0486rem + 0.4379vw, 2.4414rem)",
      variable: "--typography_fluid-arial-text-h4: clamp(2.1362rem, 2.0486rem + 0.4379vw, 2.4414rem);",
    },
  ],
  [
    "typography_fluid.arial@xxl",
    {
      key: "--typography_fluid-arial-text-xxl",
      value: "clamp(1.709rem, 1.6389rem + 0.3503vw, 1.9531rem)",
      variable: "--typography_fluid-arial-text-xxl: clamp(1.709rem, 1.6389rem + 0.3503vw, 1.9531rem);",
    },
  ],
  [
    "typography_fluid.arial@xl",
    {
      key: "--typography_fluid-arial-text-xl",
      value: "clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem)",
      variable: "--typography_fluid-arial-text-xl: clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem);",
    },
  ],
  [
    "typography_fluid.arial@l",
    {
      key: "--typography_fluid-arial-text-l",
      value: "clamp(1.0938rem, 1.0489rem + 0.2242vw, 1.25rem)",
      variable: "--typography_fluid-arial-text-l: clamp(1.0938rem, 1.0489rem + 0.2242vw, 1.25rem);",
    },
  ],
  [
    "typography_fluid.arial@m",
    {
      key: "--typography_fluid-arial-text-m",
      value: "clamp(0.875rem, 0.8391rem + 0.1794vw, 1rem)",
      variable: "--typography_fluid-arial-text-m: clamp(0.875rem, 0.8391rem + 0.1794vw, 1rem);",
    },
  ],
  [
    "typography_fluid.arial@s",
    {
      key: "--typography_fluid-arial-text-s",
      value: "clamp(0.7rem, 0.6713rem + 0.1435vw, 0.8rem)",
      variable: "--typography_fluid-arial-text-s: clamp(0.7rem, 0.6713rem + 0.1435vw, 0.8rem);",
    },
  ],
  [
    "typography_fluid.arial@xs",
    {
      key: "--typography_fluid-arial-text-xs",
      value: "clamp(0.56rem, 0.537rem + 0.1148vw, 0.64rem)",
      variable: "--typography_fluid-arial-text-xs: clamp(0.56rem, 0.537rem + 0.1148vw, 0.64rem);",
    },
  ],
]
`;

snapshot[`processTypography - can process weights 1`] = `
{
  root: "--typography_fluid-arial-2xl: clamp(1.709rem, 1.6389rem + 0.3503vw, 1.9531rem);
--typography_fluid-arial-xl: clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem);
--typography_fluid-arial-l: clamp(1.0938rem, 1.0489rem + 0.2242vw, 1.25rem);
--typography_fluid-arial-m: clamp(0.875rem, 0.8391rem + 0.1794vw, 1rem);
--typography_fluid-arial-s: clamp(0.7rem, 0.6713rem + 0.1435vw, 0.8rem);
--typography_fluid-arial-xs: clamp(0.56rem, 0.537rem + 0.1148vw, 0.64rem);
--typography_fluid-arial-2xs: clamp(0.448rem, 0.4296rem + 0.0918vw, 0.512rem);
--typography-weight-arial-regular: 500;",
}
`;

snapshot[`processTypography - can process weights 2`] = `
[
  [
    "typography_fluid.arial@2xl",
    {
      key: "--typography_fluid-arial-2xl",
      value: "clamp(1.709rem, 1.6389rem + 0.3503vw, 1.9531rem)",
      variable: "--typography_fluid-arial-2xl: clamp(1.709rem, 1.6389rem + 0.3503vw, 1.9531rem);",
    },
  ],
  [
    "typography_fluid.arial@xl",
    {
      key: "--typography_fluid-arial-xl",
      value: "clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem)",
      variable: "--typography_fluid-arial-xl: clamp(1.3672rem, 1.3111rem + 0.2803vw, 1.5625rem);",
    },
  ],
  [
    "typography_fluid.arial@l",
    {
      key: "--typography_fluid-arial-l",
      value: "clamp(1.0938rem, 1.0489rem + 0.2242vw, 1.25rem)",
      variable: "--typography_fluid-arial-l: clamp(1.0938rem, 1.0489rem + 0.2242vw, 1.25rem);",
    },
  ],
  [
    "typography_fluid.arial@m",
    {
      key: "--typography_fluid-arial-m",
      value: "clamp(0.875rem, 0.8391rem + 0.1794vw, 1rem)",
      variable: "--typography_fluid-arial-m: clamp(0.875rem, 0.8391rem + 0.1794vw, 1rem);",
    },
  ],
  [
    "typography_fluid.arial@s",
    {
      key: "--typography_fluid-arial-s",
      value: "clamp(0.7rem, 0.6713rem + 0.1435vw, 0.8rem)",
      variable: "--typography_fluid-arial-s: clamp(0.7rem, 0.6713rem + 0.1435vw, 0.8rem);",
    },
  ],
  [
    "typography_fluid.arial@xs",
    {
      key: "--typography_fluid-arial-xs",
      value: "clamp(0.56rem, 0.537rem + 0.1148vw, 0.64rem)",
      variable: "--typography_fluid-arial-xs: clamp(0.56rem, 0.537rem + 0.1148vw, 0.64rem);",
    },
  ],
  [
    "typography_fluid.arial@2xs",
    {
      key: "--typography_fluid-arial-2xs",
      value: "clamp(0.448rem, 0.4296rem + 0.0918vw, 0.512rem)",
      variable: "--typography_fluid-arial-2xs: clamp(0.448rem, 0.4296rem + 0.0918vw, 0.512rem);",
    },
  ],
  [
    "typography.weight.arial.regular",
    {
      key: "--typography-weight-arial-regular",
      value: "500",
      variable: "--typography-weight-arial-regular: 500;",
    },
  ],
]
`;
